service/dbfs/utilities.go

```diff
@@ -11,116 +11,261 @@ import (
+}
+
+// Internal only state for a write handle.
+type dbfsHandleWriter struct {
+	handle int64
+}
+
+// Internal only state for a DBFS file handle.
+type dbfsHandle struct {
+	ctx  context.Context
+	api  DbfsAPI
+	path string
+
+	reader *dbfsHandleReader
+	writer *dbfsHandleWriter
+}
+
+// Handle defines the interface of the object returned by [DbfsAPI.Open].
+type Handle interface {
+	io.ReadWriteCloser
+	io.WriterTo
+	io.ReaderFrom
+}
+
+// Implement the [io.Reader] interface.
+func (h *dbfsHandle) Read(p []byte) (int, error) {
+	r := h.reader
+	if r == nil {
+		return 0, fmt.Errorf("dbfs: file not open for reading")
 	}
-	defer func() {
-		cerr := a.CloseByHandle(ctx, handle.Handle)
-		if cerr != nil {
-			err = fmt.Errorf("close: %w", cerr)
+
+	var ntotal int
+	for ntotal < len(p) {
+		if r.offset >= r.size {
+			return ntotal, io.EOF
 		}
-	}()
-	buffer := make([]byte, 1e6)
-	for {
-		n, err := r.Read(buffer)
-		if err == io.EOF {
-			break
+
+		chunk := p[ntotal:]
+		if len(chunk) > maxDbfsBlockSize {
+			chunk = chunk[:maxDbfsBlockSize]
 		}
+
+		res, err := h.api.Read(h.ctx, Read{
+			Path:   h.path,
+			Length: len(chunk),
+			Offset: int(r.offset), // TODO: make int32/in64 work properly
+		})
 		if err != nil {
-			return fmt.Errorf("read: %w", err)
+			return ntotal, fmt.Errorf("dbfs: read: %w", err)
+		}
+
+		// The guard against offset >= size happens above, so this can only happen
+		// if the file is modified or truncated while reading. If this happens,
+		// the read contents will likely be corrupted, so we return an error.
+		if res.BytesRead == 0 {
+			return ntotal, fmt.Errorf("dbfs: read: unexpected EOF at offset %d (size %d)", r.offset, r.size)
 		}
-		err = a.AddBlock(ctx, AddBlock{
-			Data:   b64.EncodeToString(buffer[0:n]),
-			Handle: handle.Handle,
+
+		nread, err := base64.StdEncoding.Decode(chunk, []byte(res.Data))
+		if err != nil {
+			return ntotal, err
+		}
+
+		ntotal += nread
+		r.offset += int64(nread)
+	}
+
+	return ntotal, nil
+}
+
+// Implement the [io.Writer] interface.
+func (h *dbfsHandle) Write(p []byte) (int, error) {
+	w := h.writer
+	if w == nil {
+		return 0, fmt.Errorf("dbfs: file not open for writing")
+	}
+
+	var ntotal int
+	for ntotal < len(p) {
+		chunk := p[ntotal:]
+		if len(chunk) > maxDbfsBlockSize {
+			chunk = chunk[:maxDbfsBlockSize]
+		}
+
+		err := h.api.AddBlock(h.ctx, AddBlock{
+			Data:   base64.StdEncoding.EncodeToString(chunk),
+			Handle: w.handle,
 		})
 		if err != nil {
-			return fmt.Errorf("add block: %w", err)
+			return ntotal, fmt.Errorf("dbfs: add block: %w", err)
 		}
+
+		ntotal += len(chunk)
 	}
-	return err
+
+	return ntotal, nil
 }
 
-type FileReader struct {
-	Size   int64
-	ctx    context.Context
-	api    *DbfsAPI
-	path   string
-	offset int64
+// Implement the [io.Closer] interface.
+func (h *dbfsHandle) Close() error {
+	w := h.writer
+	if w == nil {
+		return fmt.Errorf("dbfs: file not open for writing")
+	}
+
+	err := h.api.CloseByHandle(h.ctx, w.handle)
```

Comment:
why not just directly return the error? :)

Reply:
Because of error nesting.

Reply:
Removing error nesting means it can return directly.